// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";
option optimize_for = LITE_RUNTIME;

import "components/manta/proto/common.proto";

package manta.proto;

enum Task {
  reserved 3, 4;

  TASK_UNSPECIFIED = 0;
  TASK_SETTINGS = 1;
  TASK_DIAGNOSTICS = 2;
  TASK_GENERIC = 5;
  TASK_PLANNER = 6;
}

enum Diagnostics {
  DIAGNOSTICS_UNSPECIFIED = 0;
  DIAGNOSTICS_MEMORY = 1;
  DIAGNOSTICS_CPU = 2;
  DIAGNOSTICS_BATTERY = 3;
  DIAGNOSTICS_STORAGE = 4;
}

enum SettingType {
  SETTING_TYPE_UNSPECIFIED = 0;
  SETTING_TYPE_BOOL = 1;
  SETTING_TYPE_STRING = 2;
  SETTING_TYPE_DOUBLE = 3;
  SETTING_TYPE_INTEGER = 4;
}

enum Role {
  ROLE_UNSPECIFIED = 0;
  ROLE_USER = 1;
  ROLE_ASSISTANT = 2;
}

message BatteryData {
  optional int32 battery_health = 1;
  optional int32 cycle_count = 2;
  optional string battery_time = 3;
  optional int32 battery_charge_percentage = 4;
}

message StorageData {
  optional string free_storage = 1;
  optional string total_storage = 2;
}

message CPUData {
  optional int32 cpu_usage_snapshot = 1;
  optional int32 temperature = 2;
  optional double clock_speed_ghz = 3;
}

message MemoryData {
  optional double free_ram_gb = 1;
  optional double total_ram_gb = 2;
}

message DiagnosticsData {
  optional BatteryData battery = 1;
  optional StorageData storage = 2;
  optional CPUData cpu = 3;
  optional MemoryData memory = 4;
}

message ServerConfig {
  optional string server_url = 1;
}

message SettingsValue {
  oneof settings_value {
    bool bool_val = 1;
    string text_val = 2;
    double double_val = 3;
    int32 int_val = 4;
  }
}

message Setting {
  string settings_id = 1;
  SettingsValue value = 2;
  SettingType type = 3;
}

message SettingsData {
  repeated Setting setting = 1;
}

message App {
  string id = 1;
  string name = 2;
  repeated string searchable_term = 3;
}

message AppsData {
  repeated App app = 1;
}

message Action {
  reserved 3, 4;

  optional Setting update_setting = 1;
  optional string launch_app_id = 2;
  optional bool all_done = 5;
}

message DiagnosticsRequest {
  repeated Diagnostics diagnostics = 1;
}

message SettingsDataRequest {}

message ContextRequest {
  reserved 2;

  optional DiagnosticsRequest diagnostics = 1;
  optional SettingsDataRequest settings = 3;
  optional Action action = 4;
}

message Turn {
  optional string message = 1;
  optional Role role = 2;
  repeated Action action = 3;
}

message SparkyResponse {
  optional ContextRequest context_request = 1;
  optional Turn latest_reply = 2;
}

message SparkyContextData {
  optional Task task = 1;
  optional DiagnosticsData diagnostics_data = 2;
  optional SettingsData settings_data = 3;
  repeated Turn conversation = 4;
  optional Image screenshot = 5;
  optional AppsData apps_data = 6;
  optional ServerConfig server_config = 7;
  optional string page_contents = 8;
  optional string page_url = 9;
}
